### PR TITLE
TaskQueue - ThreadSafe Dispose(Async) guard

### PR DESCRIPTION
**Pull request by amaitland**
Adds a double dispose guard to `DisposeAsync`. As proposed in https://github.com/hardkoded/puppeteer-sharp/pull/2118#issuecomment-1489622471 

Same pattern is used within the `dotnet runtime` itself. 

https://github.com/dotnet/runtime/blob/38b81ba83ab2b27a730a9524401d74043d172049/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs#L356

### CODE REVIEW

File: lib/PuppeteerSharp/Helpers/TaskQueue.cs
Comment (jnyrup) [inline lib/PuppeteerSharp/Helpers/TaskQueue.cs:24]: I haven't had a chance to test this out, but won't we we immediately return from `DisposeAsync` as `_disposed` has already been changed?

Comment (jnyrup) [reply lib/PuppeteerSharp/Helpers/TaskQueue.cs:24]: I had some time to test, and `_semaphore` is not disposed when calling `Dispose` for the reason described above.

If we decouple `Dispose` and `DisposeAsync` completely and call `SemaphoreSlim.Wait()` in `Dispose`, here are some unit tests to verify it.
With the thread safe guards, we should in addition be able to remove the try/catch of `ObjectDisposedException`.

```cs
public void Dispose()
{
    if (Interlocked.Exchange(ref _disposed, 1) != 0)
    {
        return;
    }

    _semaphore.Wait();
    _semaphore.Dispose();
}

public async ValueTask DisposeAsync()
{
    if (Interlocked.Exchange(ref _disposed, 1) != 0)
    {
        return;
    }

    await _semaphore.WaitAsync().ConfigureAwait(false);
    _semaphore.Dispose();
}
```

```cs
public class TaskQueueTests
{
    [Fact]
    public void ShouldDisposeSemaphoreWhenDisposing()
    {
        var taskQueue = new TaskQueue();
        taskQueue.Dispose();

        var semaphore = GetSemaphore(taskQueue);
        Assert.Throws<ObjectDisposedException>(() => semaphore.AvailableWaitHandle);
    }

    [Fact]
    public void ShouldNotThrowWhenDisposingMultipleTimes()
    {
        var taskQueue = new TaskQueue();
        taskQueue.Dispose();

        // Can be disposed a second time without throwing
        taskQueue.Dispose();
    }

    [Fact]
    public async Task ShouldDisposeSemaphoreWhenDisposingAsync()
    {
        var taskQueue = new TaskQueue();
        await taskQueue.DisposeAsync().ConfigureAwait(false);

        var semaphore = GetSemaphore(taskQueue);
        Assert.Throws<ObjectDisposedException>(() => semaphore.AvailableWaitHandle);
    }

    [Fact]
    public async Task ShouldNotThrowWhenDisposingMultipleTimesAsync()
    {
        var taskQueue = new TaskQueue();
        await taskQueue.DisposeAsync().ConfigureAwait(false);

        // Can be disposed a second time without throwing
        await taskQueue.DisposeAsync().ConfigureAwait(false);
    }

    private static SemaphoreSlim GetSemaphore(TaskQueue queue) =>
        (SemaphoreSlim)typeof(TaskQueue).GetField("_semaphore", BindingFlags.Instance | BindingFlags.NonPublic).GetValue(queue);
}
```

**Comment by kblok** [comment]
I'll wait for your approval here @jnyrup

File: lib/PuppeteerSharp/Helpers/TaskQueue.cs
Comment (amaitland) [reply lib/PuppeteerSharp/Helpers/TaskQueue.cs]: > I haven't had a chance to test this out, but won't we we immediately return from DisposeAsync as _disposed has already been changed?

You are absolutely correct! My mistake!

**Review by jnyrup — APPROVED** [review]

**Pull request merged**
